### ISSUE TITLE
distributed/rpc/tensorpipe/test_ddp_under_dist_autograd fails on release/1.6 branch

### ISSUE THREAD

**Issue by choidongyeon** [pytorch/pytorch#41365]
## 🐛 Bug

test_ddp_under_dist_autograd is failling on the release/1.6 branch with no additional changes.

## To Reproduce

Ran 
```
python test/distributed/rpc/tensorpipe/test_ddp_under_dist_autograd.py 
```

## Expected behavior

Passing test.

```
ERROR: test_ddp_dist_autograd_local_vs_remote_gpu (__main__.TestDdpComparisonTensorPipe)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/ec2-user/anaconda3/lib/python3.7/site-packages/torch/testing/_internal/common_distributed.py", line 204, in wrapper
    self._join_processes(fn)
  File "/home/ec2-user/anaconda3/lib/python3.7/site-packages/torch/testing/_internal/common_distributed.py", line 306, in _join_processes
    self._check_return_codes(elapsed_time)
  File "/home/ec2-user/anaconda3/lib/python3.7/site-packages/torch/testing/_internal/common_distributed.py", line 339, in _check_return_codes
    raise RuntimeError(error)
RuntimeError: Processes 4 5 exited with error code 10
```

## Environment

```
Collecting environment information...
PyTorch version: N/A
Is debug build: N/A
CUDA used to build PyTorch: N/A

OS: Amazon Linux 2
GCC version: (GCC) 7.3.1 20180712 (Red Hat 7.3.1-6)
CMake version: version 3.13.3

Python version: 3.7
Is CUDA available: N/A
CUDA runtime version: 10.0.130
GPU models and configuration: 
GPU 0: Tesla V100-SXM2-16GB
GPU 1: Tesla V100-SXM2-16GB
GPU 2: Tesla V100-SXM2-16GB
GPU 3: Tesla V100-SXM2-16GB

Nvidia driver version: 440.33.01
cuDNN version: Probably one of the following:
/usr/local/cuda-10.1/targets/x86_64-linux/lib/libcudnn.so.7.6.5
/usr/local/cuda-10.2/targets/x86_64-linux/lib/libcudnn.so.7.6.5

Versions of relevant libraries:
[pip3] numpy==1.18.5
[conda] blas                      1.0                         mkl  
[conda] mkl                       2020.0                      166  
[conda] mkl-service               2.3.0            py37he904b0f_0  
[conda] mkl_fft                   1.0.15           py37ha843d7b_0  
[conda] mkl_random                1.1.0            py37hd6b4f25_0  
[conda] numpy                     1.18.1           py37h4f9e942_0  
[conda] numpy-base                1.18.1           py37hde5b4d6_1  
[conda] numpydoc                  0.9.2                      py_0  
[conda] torch                     1.6.0a0+cefb9e0          pypi_0    pypi
```

cc @ezyang @gchanan @zou3519 @pietern @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @xush6528 @osalpekar @jiayisuse @agolynski @jjlilley @lw @beauby

**Comment by rohan-varma** [comment]
cc @pritamdamania87

**Comment by rohan-varma** [comment]
@choidongyeon I tried to repro this by running it a bunch of times, and wasn't able to so far. There was one fix submitted a few weeks back https://github.com/pytorch/pytorch/pull/40495, that had to be cherry-picked into this branch. I just wanted to confirm if your local checkout of the release branch has this cherry-pick in (if you are on the most recent `release/1.6`. branch then it should)?

**Comment by lw** [comment]
Based on the environment dump the checkout should be on commit cefb9e0 (see the last line), which is at the moment the latest one on release/1.6. So that should include the fixes.

@choidongyeon Were there other relevant logs printed above that snippet you posted? That snippet comes from the parent process reporting that some children failed, but I think the children should also have printed something to explain why they failed.

**Comment by pritamdamania87** [comment]
@choidongyeon Could you provide the entire log for the test failure? Usually there should be some additional log above what you pasted which logs the error that was actually encountered on process 4 and 5.

**Comment by Flamefire** [comment]
I encountered the same and found the issue: The problem occurs when there are less than 6 GPUs but at least 4 and is a wrong test annotation.

From the log:
```
Traceback (most recent call last):
  File "/tmp/easybuild-tmp/eb-KvsELm/tmpUgp1Un/lib/python3.7/site-packages/torch/testing/_internal/common_distributed.py", line 207, in wrapper
    fn()
  File "/tmp/easybuild-tmp/eb-KvsELm/tmpUgp1Un/lib/python3.7/site-packages/torch/testing/_internal/common_distributed.py", line 76, in wrapper
    return func(*args, **kwargs)
  File "/tmp/easybuild-tmp/eb-KvsELm/tmpUgp1Un/lib/python3.7/site-packages/torch/testing/_internal/dist_utils.py", line 93, in new_test_method
    return_value = old_test_method(self, *arg, **kwargs)
  File "/tmp/easybuild-tmp/eb-KvsELm/tmpUgp1Un/lib/python3.7/site-packages/torch/testing/_internal/distributed/ddp_under_dist_autograd_test.py", line 616, in test_ddp_dist_autograd_local_vs_remote_gpu
    layer2 = nn.Linear(7, 5).cuda(self.rank)
  File "/tmp/easybuild-tmp/eb-KvsELm/tmpUgp1Un/lib/python3.7/site-packages/torch/nn/modules/module.py", line 458, in cuda
    return self._apply(lambda t: t.cuda(device))
  File "/tmp/easybuild-tmp/eb-KvsELm/tmpUgp1Un/lib/python3.7/site-packages/torch/nn/modules/module.py", line 376, in _apply
    param_applied = fn(param)
  File "/tmp/easybuild-tmp/eb-KvsELm/tmpUgp1Un/lib/python3.7/site-packages/torch/nn/modules/module.py", line 458, in <lambda>
    return self._apply(lambda t: t.cuda(device))
RuntimeError: CUDA error: invalid device ordinal
```

Now checking `pytorch/test/distributed/rpc/tensorpipe/test_ddp_under_dist_autograd.py` reveals: 

```
    @property
    def world_size(self) -> int:
        return ddp_under_dist_autograd_test.WORLD_SIZE
```

This is set as:

```
NUM_TRAINERS = 4
# Trainers + the master + the remote worker
WORLD_SIZE = NUM_TRAINERS + 2
```

However the test is annotated with:

```
    @skip_if_lt_x_gpu(NUM_TRAINERS)
    @requires_nccl()
    @dist_init
    def test_ddp_dist_autograd_local_vs_remote_gpu(self):
```

And it calls `layer2 = nn.Linear(7, 5).cuda(self.rank)`.

So in summary: 6 ranks are started and it tries to get 1 GPU for each rank with the index being the rank index. Hence rank 5 & 6 will fail as there is no such GPU

**Comment by Flamefire** [comment]
Or the world_size is wrong. Because the test class is defined via inheritance  as `class class TestDdpComparisonTensorPipe(TensorPipeRpcAgentTestFixture, ddp_under_dist_autograd_test.TestDdpComparison):` but the original test from TestDdpComparison uses `NUM_TRAINERS` as the world size: https://github.com/pytorch/pytorch/blob/cb9029df9db397f66da234d21b5ddafd6be6e602/torch/testing/_internal/distributed/ddp_under_dist_autograd_test.py#L438

**Comment by rgommers** [comment]
This was fixed by gh-41995, so closing. Your diagnosis of the issue was correct @Flamefire, thanks!   And thanks @choidongyeon for reporting.

The fix was included in 1.7.0 The test file changed too, it's now:
```
python test/distributed/rpc/test_tensorpipe_agent.py
```
